src/main/java/io/knowledgeassets/myskills/server/community/command/CommunityUserCommandService.java

```diff
@@ -0,0 +1,95 @@
+import io.knowledgeassets.myskills.server.user.User;
+import io.knowledgeassets.myskills.server.user.query.UserQueryService;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+
+@Service
+public class CommunityUserCommandService {
+
+	private final CommunityRepository communityRepository;
+	private final UserQueryService userQueryService;
+
+	public CommunityUserCommandService(CommunityRepository communityRepository,
+								   UserQueryService userQueryService) {
+		this.communityRepository = communityRepository;
+		this.userQueryService = userQueryService;
+	}
+
+	/**
+	 * User joins the community as member.
+	 * @param communityId - id of the community to join
+	 * @param userId - user id
+	 */
+	@Transactional
+	public Community joinCommunityAsMember(String communityId, String userId) {
+		final User user = userQueryService.getUserById(userId)
+				.orElseThrow(buildNoSuchResourceExceptionSupplier("id", userId, Model.USER));
+		final Community community = communityRepository.findById(communityId).orElseThrow(buildNoSuchResourceExceptionSupplier("communityId", communityId, Model.COMMUNITY));
+		if (CommunityType.OPENED.equals(community.getType())) {
+			final List<User> members;
+			if (community.getMembers() != null) {
+				members = new ArrayList<>(community.getMembers());
+			} else {
+				members = new ArrayList<>();
+			}
+			members.add(user);
+			community.setMembers(members);
+			community.setLastModifiedDate(LocalDateTime.now());
+			return communityRepository.save(community);
+		}
+		else {
+			throw new CommunityAccessDeniedException(format("The community with ID \"%s\" is not an open one." +
+					" Only open communities can be joined without request.", communityId));
+		}
+	}
+
+	private Supplier<NoSuchResourceException> buildNoSuchResourceExceptionSupplier(String paramName, String paramValue, Model community) {
+		return () -> {
+			final String[] searchParamsMap = {paramName, paramValue};
+			return NoSuchResourceException.builder()
+					.model(community)
+					.searchParamsMap(searchParamsMap)
+					.build();
+		};
+	}
+
+	/**
+	 * Authenticated user leaves the community.
+	 * @param communityId - id of the community to leave
+	 * @param userId - user id
+	 */
+	@Transactional
+	public Community leaveCommunity(String communityId, String userId) {
+		final User user = userQueryService.getUserById(userId)
+				.orElseThrow(buildNoSuchResourceExceptionSupplier("id", userId, Model.USER));
+		final Community community = communityRepository.findById(communityId).orElseThrow(buildNoSuchResourceExceptionSupplier("communityId", communityId, Model.COMMUNITY));
+
+		if (community.getMembers().stream().noneMatch((User u) -> user.getId().equals(u.getId()))) {
```

Comment:
Good point. To tell the truth the check above was unnecessary as it is always guaranteed by
[CommunityCommandService#create](https://github.com/KnowledgeAssets/myskills-server/blob/master/src/main/java/io/knowledgeassets/myskills/server/community/command/CommunityCommandService.java#L46) there is at least one member of a community (its manager) and the collection of members cannot be empty. The member / manager collections can never be empty as there has to be at least one manager of a community and she / he will be a member either (in the case of an open community a community manager will be its creator). There is a requirement from the PO that `In both types of communities the role "communitiy manager" can change the role of other users but not his/her own`. As a result there has to be at least one community manager as she / he cannot change her / his own role.

I removed the check on `null` above to avoid any confusions.